File: hexwar/lua/tests/test_prim.lua
-- 골든 프리미티브 대조 (루아) — golden/prim.txt 를 파이썬판과 같은 순서로 읽는다.

package.path = './?.lua;' .. package.path

local H = require('hexwar.hexcoord')
local hexmap = require('hexwar.hexmap')
local PK = require('hexwar.picker')
local R = require('hexwar.rng')

local fails = {}

local function eq(name, got, want)
  if got ~= want then
    fails[#fails + 1] = string.format('%s: got %s want %s', name, tostring(got), tostring(want))
  end
end

-- got 을 want[from..] 과 견준다. n 을 주면 그만큼만, 안 주면 끝까지 같아야 한다.
local function eqlist(name, got, want, from, n)
  from = from or 1
  if n == nil and #got ~= (#want - from + 1) then
    fails[#fails + 1] = string.format('%s: 길이 %d != %d', name, #got, #want - from + 1)
    return
  end
  if n ~= nil and #got ~= n then
    fails[#fails + 1] = string.format('%s: 길이 %d != %d', name, #got, n)
    return
  end
  for i = 1, #got do
    if got[i] ~= want[from + i - 1] then
      fails[#fails + 1] = string.format('%s: [%d] %d != %d', name, i, got[i], want[from + i - 1])
      return
    end
  end
end

local function unhex(s)
  if s == '-' then return '' end
  local out = {}
  for i = 1, #s, 2 do out[#out + 1] = string.char(tonumber(s:sub(i, i + 1), 16)) end
  return table.concat(out)
end

local path = (os.getenv('HEXWAR_GOLDEN') or '../golden') .. '/prim.txt'
local nline = 0
for line in io.lines(path) do
  if line ~= '' and line:sub(1, 1) ~= ';' then
    nline = nline + 1
    local parts = {}
    for w in line:gmatch('%S+') do parts[#parts + 1] = w end
    local key = parts[1]
    if key == 'fnv' then
      eq('fnv ' .. parts[2], R.fnv1a(unhex(parts[2])), tonumber(parts[3]))
    else
      local v = {}
      for i = 2, #parts do v[i - 1] = tonumber(parts[i]) end
      if key == 'dirs' then
        local got = {}
        for d = 0, 5 do got[#got + 1] = H.DIRS[d][1]; got[#got + 1] = H.DIRS[d][2] end
        eqlist('dirs', got, v)
      elseif key == 'oddr' then
        local c, r = H.axial_to_oddr(v[1], v[2])
        eqlist('axial_to_oddr', { c, r }, v, 3)
        local q, rr = H.oddr_to_axial(v[3], v[4])
        eqlist('oddr_to_axial', { q, rr }, v, 1, 2)
      elseif key == 'oddq' then
        local c, r = H.axial_to_oddq(v[1], v[2])
        eqlist('axial_to_oddq', { c, r }, v, 3)
        local q, rr = H.oddq_to_axial(v[3], v[4])
        eqlist('oddq_to_axial', { q, rr }, v, 1, 2)
      elseif key == 'dist' then
        eq('distance', H.distance(v[1], v[2], v[3], v[4]), v[5])
      elseif key == 'neighbors' then
        local got = {}
        for _, h in ipairs(H.neighbors(v[1], v[2])) do
          got[#got + 1] = h[1]; got[#got + 1] = h[2]
        end
        eqlist('neighbors', got, v, 3)
      elseif key == 'ring' then
        local got = {}
        for _, h in ipairs(H.ring(0, 0, v[1])) do got[#got + 1] = h[1]; got[#got + 1] = h[2] end
        eqlist('ring' .. v[1], got, v, 2)
      elseif key == 'spiral' then
        eq('spiral' .. v[1], #H.spiral(0, 0, v[1]), v[2])
      elseif key == 'line' then
        local hexes = H.line(v[1], v[2], v[3], v[4])
        local got = { #hexes }
        for _, h in ipairs(hexes) do got[#got + 1] = h[1]; got[#got + 1] = h[2] end
        eqlist('line', got, v, 5)
      elseif key == 'pick' then
        local col, row = PK.pick(v[1], v[2], v[3], v[4])
        eqlist('pick', { col or -1, row or -1 }, v, 5)
      elseif key == 'lcg' then
        local st = R.new(v[1])
        local got = {}
        for i = 2, #v do got[#got + 1] = st:next() end
        eqlist('lcg', got, v, 2)
      elseif key == 'd6' then
        local st = R.new(0x1BADB002)
        local got = {}
        for i = 1, #v do got[#got + 1] = st:d6() end
        eqlist('d6', got, v)
      elseif key == 'cell' then
        eq('pack', hexmap.pack_cell(v[1], v[2], v[3]), v[4])
        eqlist('unpack', { hexmap.cell_terrain(v[4]), hexmap.cell_elev(v[4]),
                           hexmap.cell_road(v[4]) }, v, 1, 3)
      else
        fails[#fails + 1] = '알 수 없는 키: ' .. key
      end
    end
  end
end

-- 마스크 표가 golden/pick_mask.txt 와 같은지도 본다
do
  local oy = 0
  for line in io.lines((os.getenv('HEXWAR_GOLDEN') or '../golden') .. '/pick_mask.txt') do
    if line ~= '' then
      for ox = 0, 31 do
        eq(string.format('mask[%d][%d]', oy, ox),
           PK.PICK_MASK[oy * 32 + ox], tonumber(line:sub(ox + 1, ox + 1)))
      end
      oy = oy + 1
    end
  end
  eq('mask 행 수', oy, 24)
end

if #fails > 0 then
  print(string.format('FAIL %d / %d줄', #fails, nline))
  for i = 1, math.min(20, #fails) do print('  ' .. fails[i]) end
  os.exit(1)
end
print(string.format('prim OK (lua) — %d줄 + 마스크 768칸 전부 일치', nline))

File: hexwar/ts/tests/test_prim.ts
// 골든 프리미티브 대조 (타입스크립트) — 파이썬·루아판과 같은 파일, 같은 순서.

import * as fs from 'fs';
import * as path from 'path';

import * as H from '../src/hexcoord';
import * as M from '../src/hexmap';
import * as PK from '../src/picker';
import { Rng, fnv1a } from '../src/rng';

const GOLDEN = process.env['HEXWAR_GOLDEN'] ?? path.join(__dirname, '..', '..', '..', 'golden');
const fails: string[] = [];

function eq(name: string, got: unknown, want: unknown): void {
  if (got !== want) fails.push(`${name}: got ${String(got)} want ${String(want)}`);
}

function eqList(name: string, got: number[], want: number[], from = 0, n?: number): void {
  const count = n ?? want.length - from;
  if (got.length !== count) {
    fails.push(`${name}: 길이 ${got.length} != ${count}`);
    return;
  }
  for (let i = 0; i < got.length; i++) {
    if (got[i] !== want[from + i]) {
      fails.push(`${name}: [${i}] ${got[i]} != ${want[from + i]}`);
      return;
    }
  }
}

function unhex(s: string): Uint8Array {
  if (s === '-') return new Uint8Array(0);
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(s.slice(i * 2, i * 2 + 2), 16);
  return out;
}

let nline = 0;
for (const raw of fs.readFileSync(path.join(GOLDEN, 'prim.txt'), 'utf8').split('\n')) {
  const line = raw.trim();
  if (!line || line.startsWith(';')) continue;
  nline++;
  const parts = line.split(/\s+/);
  const key = parts[0]!;
  if (key === 'fnv') {
    eq(`fnv ${parts[1]}`, fnv1a(unhex(parts[1]!)), Number(parts[2]));
    continue;
  }
  const v = parts.slice(1).map(Number);
  switch (key) {
    case 'dirs':
      eqList('dirs', H.DIRS.flatMap((d) => [d[0], d[1]]), v);
      break;
    case 'oddr':
      eqList('axialToOddr', [...H.axialToOddr(v[0]!, v[1]!)], v, 2);
      eqList('oddrToAxial', [...H.oddrToAxial(v[2]!, v[3]!)], v, 0, 2);
      break;
    case 'oddq':
      eqList('axialToOddq', [...H.axialToOddq(v[0]!, v[1]!)], v, 2);
      eqList('oddqToAxial', [...H.oddqToAxial(v[2]!, v[3]!)], v, 0, 2);
      break;
    case 'dist':
      eq('distance', H.distance(v[0]!, v[1]!, v[2]!, v[3]!), v[4]);
      break;
    case 'neighbors':
      eqList('neighbors', H.neighbors(v[0]!, v[1]!).flatMap((h) => [h[0], h[1]]), v, 2);
      break;
    case 'ring':
      eqList(`ring${v[0]}`, H.ring(0, 0, v[0]!).flatMap((h) => [h[0], h[1]]), v, 1);
      break;
    case 'spiral':
      eq(`spiral${v[0]}`, H.spiral(0, 0, v[0]!).length, v[1]);
      break;
    case 'line': {
      const hexes = H.line(v[0]!, v[1]!, v[2]!, v[3]!);
      eqList('line', [hexes.length, ...hexes.flatMap((h) => [h[0], h[1]])], v, 4);
      break;
    }
    case 'pick': {
      const p = PK.pick(v[0]!, v[1]!, v[2]!, v[3]!);
      eqList('pick', p ? [p[0], p[1]] : [-1, -1], v, 4);
      break;
    }
    case 'lcg': {
      const st = new Rng(v[0]!);
      eqList('lcg', v.slice(1).map(() => st.next()), v, 1);
      break;
    }
    case 'd6': {
      const st = new Rng(0x1badb002);
      eqList('d6', v.map(() => st.d6()), v);
      break;
    }
    case 'cell':
      eq('pack', M.packCell(v[0]!, v[1]!, v[2]!), v[3]);
      eqList('unpack', [M.cellTerrain(v[3]!), M.cellElev(v[3]!), M.cellRoad(v[3]!)], v, 0, 3);
      break;
    default:
      fails.push(`알 수 없는 키: ${key}`);
  }
}

// 마스크 표가 golden/pick_mask.txt 와 같은지도 본다
{
  const lines = fs.readFileSync(path.join(GOLDEN, 'pick_mask.txt'), 'utf8')
    .split('\n').filter((s) => s !== '');
  eq('mask 행 수', lines.length, 24);
  for (let oy = 0; oy < lines.length; oy++) {
    for (let ox = 0; ox < 32; ox++) {
      eq(`mask[${oy}][${ox}]`, PK.PICK_MASK[oy * 32 + ox], Number(lines[oy]![ox]));
    }
  }
}

if (fails.length > 0) {
  console.log(`FAIL ${fails.length} / ${nline}줄`);
  for (const f of fails.slice(0, 20)) console.log('  ' + f);
  process.exit(1);
}
console.log(`prim OK (ts) — ${nline}줄 + 마스크 768칸 전부 일치`);
